cpp_modules/src/mcts/Node.h

```diff
@@ -8,61 +8,65 @@
 #include <map>
 #include <memory>
 #include <utility>
-#include <ValueHeuristicEnum.h>
 #include <vector>
 
 #include "ActionEnum.h"
 
 namespace Sc2::Mcts {
 	class Node : public std::enable_shared_from_this<Node> {
-		ValueHeuristic _valueHeuristic;
-
-		Action action;
-		std::shared_ptr<Node> parent;
+		Action _action;
 		int depth = 0;
 
+		std::shared_ptr<Node> _parent;
+
+		std::shared_ptr<State> _state;
+
 	public:
 		// Number of simulations that has been run on this node
 		int N = 0;
 		// The cumulative value of this node
 		double Q = 0;
 		std::map<Action, std::shared_ptr<Node> > children = {};
 
-		std::shared_ptr<Node> getParent() { return parent; }
-		[[nodiscard]] Action getAction() const { return action; }
+		std::shared_ptr<State> getState() { return _state; }
+		void setParent(std::shared_ptr<Node> parent) { _parent = std::move(parent); }
+		std::shared_ptr<Node> getParent() { return _parent; }
+		[[nodiscard]] Action getAction() const { return _action; }
 		int getDepth() const { return depth; }
 
-		void expand(const std::shared_ptr<State> &state) {
-			const std::vector<Action> actions = state->getLegalActions();
+		void expand() {
+			const std::vector<Action> actions = _state->getLegalActions();
 			addChildren(actions);
 		}
 
 		void addChildren(const std::vector<Action> &childActions) {
 			for (const auto &childAction: childActions) {
-				const auto childNode = std::make_shared<Node>(Node(childAction, shared_from_this(), _valueHeuristic));;
+				const auto state = State::DeepCopy(*_state);
+
+				const auto childNode = std::make_shared<Node>(Node(childAction, shared_from_this(), state));
 				childNode->depth = this->depth + 1;
-				children[childNode->action] = childNode;
+				children[childNode->_action] = childNode;
 			}
 		}
 
 		[[nodiscard]] std::string toString() const {
```

Comment:
Nothing bad about this. Just wanted to say that the toString method looks really clean, and it's still very nice to read 👍